tests/Datagrid/DatagridTest.php

```diff
@@ -62,7 +62,6 @@ final class DatagridTest extends TestCase
     protected function setUp(): void
     {
         $this->query = $this->createStub(ProxyQueryInterface::class);
-        \assert($this->query instanceof ProxyQueryInterface); // https://github.com/vimeo/psalm/issues/5818
```

Comment:
I added this for psalm, we can either
- Ignore the "always true" error for phpstan.
- Find another way to solve the psalm issue. 

Currently the phpdoc is 
```
    /**
     * @var Datagrid<ProxyQueryInterface>
     */
    private $datagrid;
```
So we have to pass a `ProxyQueryInterface` to the datagrid constructor and not a `ProxyQueryInterface&Stub`.

I think if you change the phpdoc to
```
    /**
     * @var Datagrid<ProxyQueryInterface&Stub>
     */
    private $datagrid;

    /**
     * @var PagerInterface<ProxyQueryInterface&Stub>&MockObject
     */
    private $pager;

    /**
     * @var ProxyQueryInterface&Stub
     */
    private $query;
```
the issue will be solved.

Reply:
Edit: it will be solved for psalm, but then phpstan throw
```
  73     Property Sonata\AdminBundle\Tests\Datagrid\DatagridTest::$datagrid (Sonata\AdminBundle\Datagrid\Datagrid<PHPUnit\Framework\MockObject\Stub&Sonata\AdminBundle\Datagrid\ProxyQueryInterface>) does not accept  
         Sonata\AdminBundle\Datagrid\Datagrid<Sonata\AdminBundle\Datagrid\ProxyQueryInterface>.                                                                                                                        
  446    Parameter #3 $pager of class Sonata\AdminBundle\Datagrid\Datagrid constructor expects Sonata\AdminBundle\Datagrid\PagerInterface<Sonata\AdminBundle\Datagrid\ProxyQueryInterface>,                            
         PHPUnit\Framework\MockObject\MockObject&Sonata\AdminBundle\Datagrid\PagerInterface<PHPUnit\Framework\MockObject\Stub&Sonata\AdminBundle\Datagrid\ProxyQueryInterface> given.                                  
  446    Property Sonata\AdminBundle\Tests\Datagrid\DatagridTest::$datagrid (Sonata\AdminBundle\Datagrid\Datagrid<PHPUnit\Framework\MockObject\Stub&Sonata\AdminBundle\Datagrid\ProxyQueryInterface>) does not accept  
         Sonata\AdminBundle\Datagrid\Datagrid<Sonata\AdminBundle\Datagrid\ProxyQueryInterface>.                                                                                                                        
  496    Parameter #3 $pager of class Sonata\AdminBundle\Datagrid\Datagrid constructor expects Sonata\AdminBundle\Datagrid\PagerInterface<Sonata\AdminBundle\Datagrid\ProxyQueryInterface>,                            
         PHPUnit\Framework\MockObject\MockObject&Sonata\AdminBundle\Datagrid\PagerInterface<PHPUnit\Framework\MockObject\Stub&Sonata\AdminBundle\Datagrid\ProxyQueryInterface> given.                                  
  496    Property Sonata\AdminBundle\Tests\Datagrid\DatagridTest::$datagrid (Sonata\AdminBundle\Datagrid\Datagrid<PHPUnit\Framework\MockObject\Stub&Sonata\AdminBundle\Datagrid\ProxyQueryInterface>) does not accept  
         Sonata\AdminBundle\Datagrid\Datagrid<Sonata\AdminBundle\Datagrid\ProxyQueryInterface>.                                                                                                                        
  555    Parameter #3 $pager of class Sonata\AdminBundle\Datagrid\Datagrid constructor expects Sonata\AdminBundle\Datagrid\PagerInterface<Sonata\AdminBundle\Datagrid\ProxyQueryInterface>,                            
         PHPUnit\Framework\MockObject\MockObject&Sonata\AdminBundle\Datagrid\PagerInterface<PHPUnit\Framework\MockObject\Stub&Sonata\AdminBundle\Datagrid\ProxyQueryInterface> given.                                  
  555    Property Sonata\AdminBundle\Tests\Datagrid\DatagridTest::$datagrid (Sonata\AdminBundle\Datagrid\Datagrid<PHPUnit\Framework\MockObject\Stub&Sonata\AdminBundle\Datagrid\ProxyQueryInterface>) does not accept  
         Sonata\AdminBundle\Datagrid\Datagrid<Sonata\AdminBundle\Datagrid\ProxyQueryInterface>.                                                                                                                        
  593    Property Sonata\AdminBundle\Tests\Datagrid\DatagridTest::$datagrid (Sonata\AdminBundle\Datagrid\Datagrid<PHPUnit\Framework\MockObject\Stub&Sonata\AdminBundle\Datagrid\ProxyQueryInterface>) does not accept  
         Sonata\AdminBundle\Datagrid\Datagrid<Sonata\AdminBundle\Datagrid\ProxyQueryInterface>.                                                                                                                        
  596    Parameter #3 $pager of class Sonata\AdminBundle\Datagrid\Datagrid constructor expects Sonata\AdminBundle\Datagrid\PagerInterface<Sonata\AdminBundle\Datagrid\ProxyQueryInterface>,                            
         PHPUnit\Framework\MockObject\MockObject&Sonata\AdminBundle\Datagrid\PagerInterface<PHPUnit\Framework\MockObject\Stub&Sonata\AdminBundle\Datagrid\ProxyQueryInterface> given.                                  
  616    Property Sonata\AdminBundle\Tests\Datagrid\DatagridTest::$datagrid (Sonata\AdminBundle\Datagrid\Datagrid<PHPUnit\Framework\MockObject\Stub&Sonata\AdminBundle\Datagrid\ProxyQueryInterface>) does not accept  
         Sonata\AdminBundle\Datagrid\Datagrid<Sonata\AdminBundle\Datagrid\ProxyQueryInterface>.                                                                                                                        
  619    Parameter #3 $pager of class Sonata\AdminBundle\Datagrid\Datagrid constructor expects Sonata\AdminBundle\Datagrid\PagerInterface<Sonata\AdminBundle\Datagrid\ProxyQueryInterface>,                            
         PHPUnit\Framework\MockObject\MockObject&Sonata\AdminBundle\Datagrid\PagerInterface<PHPUnit\Framework\MockObject\Stub&Sonata\AdminBundle\Datagrid\ProxyQueryInterface> given.                                  
  634    Property Sonata\AdminBundle\Tests\Datagrid\DatagridTest::$datagrid (Sonata\AdminBundle\Datagrid\Datagrid<PHPUnit\Framework\MockObject\Stub&Sonata\AdminBundle\Datagrid\ProxyQueryInterface>) does not accept  
         Sonata\AdminBundle\Datagrid\Datagrid<Sonata\AdminBundle\Datagrid\ProxyQueryInterface>.                                                                                                                        
  637    Parameter #3 $pager of class Sonata\AdminBundle\Datagrid\Datagrid constructor expects Sonata\AdminBundle\Datagrid\PagerInterface<Sonata\AdminBundle\Datagrid\ProxyQueryInterface>,                            
         PHPUnit\Framework\MockObject\MockObject&Sonata\AdminBundle\Datagrid\PagerInterface<PHPUnit\Framework\MockObject\Stub&Sonata\AdminBundle\Datagrid\ProxyQueryInterface> given. 
```

This need more understanding.

Reply:
But I think that's the right way to solve it and created a phpstan issue https://github.com/phpstan/phpstan/issues/5336

Reply:
I just reverted this change, it can be tackled later. As you said I tried your proposal but phpstan fails.